substrate/client/network/src/protocol/notifications/handler.rs

```diff
@@ -680,17 +682,19 @@ impl ConnectionHandler for NotifsHandler {
 
 		// A grace period of `INITIAL_KEEPALIVE_TIME` must be given to leave time for the remote
 		// to express desire to open substreams.
+		#[allow(deprecated)]
 		KeepAlive::Until(self.when_connection_open + INITIAL_KEEPALIVE_TIME)
 	}
 
+	#[allow(deprecated)]
 	fn poll(
```

Comment:
I'm out of context here, but do we have a plan to mitigate using the deprecated functions?
Maybe we can do a follow-up after merging this? Or switch entirely to litep2p soon?

Reply:
While I'm curious about the use of deprecated functions as well, I also want to ask about the "litep2p" mention. While it's by a polkadot-sdk contributor, it hardly seems ready for the discussion to replace libp2p. Is that the long term goal of LiteP2P? Is there any public conversations I can read about this?

Reply:
I haven't made a definitive decision regarding these deprecation notices. This upgrade has been an absolute nightmare to get working and as you can see from the failing zombienet tests, it's still not ready. In addition to this PR, I have other things to work on as well so I'm not sure how much time I want to spend on rewriting code that already works. They will most likely be fixed at some point.

As far as litep2p goes, the plan is not to replace rust-libp2p, at least not yet. I have a WIP branch I've been working on and it will be merged at some point. After the branch is merged, rust-libp2p will still be the default networking backend but users can choose to run Polkadot SDK with litep2p. Long-term goal is to make litep2p the default backend and potentially phase out rust-libp2p, time will tell.

There is no public discussion about it yet but once the PR that integrates litep2p into Polkadot SDK has been opened/merged, I will make a post about it on the forum.

Reply:
I definitely get this PR has been a massive effort, which I appreciate and look forward to (mainly due to the cryptographic dependencies being updated).

Thanks for the initial comment on litep2p. While incredibly skeptical, I'll wait and look forward to your post :)

Reply:
`rust-libp2p` maintainer here.

> This upgrade has been an absolute nightmare to get working

I hear your frustration and I just want to say that we actually put what I'd consider to be a lot of effort into only making breaking changes when we deem them to be a net-positive. Historically, `rust-libp2p` has an extensive public API surface which unfortunately means that even tiny changes some parts are backwards-incompatible :(

For example, one of the larger breaking changes in `v0.52` that would affect users that author their own protocols was https://github.com/libp2p/rust-libp2p/pull/3605. I believe it s a good change because it simplifies error handling and removes a lot of "what am I meant to do about this error?" questions.

Similarly, and this is what the `allow(deprecated)` of this thread silences, in `v0.53`, we've greatly simplified connection keep-alive and closing. This is now just built-in and you no longer need to deal with it[^1]. We also removed `ConnectionHandlerEvent::Close` which a) removes tons of type-parameters and thus simplifies the API and b) improves resilience because protocols aren't simply aborted mid-operation. We've extensively discussed this in https://github.com/libp2p/rust-libp2p/discussions/3353 if you are interested on the details :)

My primary personal maintainer goal over the last years has been to _improve_ ergonomics so using `rust-libp2p` is _easier_ not harder and I think we also made good progress on this front. However, I am also cognizant that without being involved in a downstream project on a daily basis, how well something is working for our users can be a bit hard to gauge. We are definitely open to feedback here so we can make better decisions on what should be prioritized.

> Long-term goal is to make litep2p the default backend and potentially phase out rust-libp2p, time will tell.

Whilst you are obviously free to do whatever you want, I think there would be lots of value in collaborating on making `rust-libp2p` work well for you so you don't have to go through all the effort of rewriting it.

[^1]: Unless you have a special case like the relay protocols where even a connection without streams should not be killed. You can read more about this in the release notes: https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-v0.53.0

Reply:
I appreciate the effort of making rust-libp2p more ergonomic and I hope we're able to reap those benefits as well. Maybe the challenge for us is that our networking stack works and has been working for a long time so it doesn't need that much maintaining (as long as it's working). But if we want to get a stable version of QUIC, we have to rewrite parts of our notification protocol or update the way we deal with external addresses, it's a lot of work in unrelated areas and the costs for getting all of the changes to a state where we feel confident in releasing them is high.

For example, this upgrade is needed because we need to update schnorrkel and while optimally that would be a simple dependency update in our end, it requires quite a bit more work. It could also be a patch release to 0.51 but whether that's even possible or something you're willing to entertain is another matter. But right now it means we have to update and fix everything that changed in the release just to get WebRTC updated. 

>Whilst you are obviously free to do whatever you want, I think there would be lots of value in collaborating on making rust-libp2p work well for you so you don't have to go through all the effort of rewriting it. At the end of the day, polkadot is one of the largest users of rust-libp2p so I believe it should be working well for you! :)

I think there is plenty of room for collaboration, irrespective of which direction is chosen. If we decide to stay with rust-libp2p, then we'll be contributing upstream and if we decide to focus on litep2p, we have more freedom to experiment and maybe some of the results are something that would benefit the libp2p collective at large and we can discuss if it could be a standard libp2p protocol instead. Also, the lower parts of the stack (yamux, noise, TLS, etc.) are areas where we can potentially share code more freely.

Reply:
> It could also be a patch release to 0.51 but whether that's even possible or something you're willing to entertain is another matter.

In general, we are very open to backports. We maintain a long-lived branch for each minor release that you can open a PR against if you would like to backport something.

> we have to rewrite parts of our notification protocol

I'd be curious to learn what was particularly difficult about that so we can make better decisions in the future. From quickly glancing over the diff, it looks quite noisy but most of it is renames of associated types?

The tests haven't been ported yet. I am not sure if that is something you want to do as part of the upgrade but what might come in handy is our `libp2p-swarm-test` crate. In case you don't know about it, it allows you to easily create instances of `Swarm` that are connected via `/memory` or `/tcp` and have lots of utility functions for connecting them and / or waiting for events. We use it across our entire codebase to test all `NetworkBehaviour`s and introducing it has massively cut down the boilerplate for writing tests.

Additionally, it means you are testing the public API of your `NetworkBehaviour` and not internals like handlers or stream upgrades. See https://github.com/libp2p/rust-libp2p/blob/3ec575aad08558911f5056326e5a5fad83830b82/protocols/identify/tests/smoke.rs#L87-L157 for an example.

In general, I'd have two recommendations to make future upgrades easier:

- Try to expose yourself as little as possible to the APIs of `NetworkBehaviour` and `ConnectionHandler`. Don't write (many) tests against it. Those APIs should only be called by the `Swarm`. Most of the breaking changes we decide to make concern these interfaces and are born out of a need for either of those components to react to state updates and / or send new events. Implementations of these traits will be hit by the breaking changes anyway, that is unavoidable. But if your tests are at least written against an instance of `Swarm`, those shouldn't need touching which means a) less work and b) gives you more confidence that everything works if they are still passing after the update.
- Do not sit on `allow(deprecated)` for too long as it will create a time-lag in when you can update to the next version. In particular, `v0.52` was already released 4 months ago and we've actually just released [`v0.53`](https://github.com/libp2p/rust-libp2p/releases/tag/libp2p-v0.53.0) last week.

Reply:
>I'd be curious to learn what was particularly difficult about that so we can make better decisions in the future. 

I haven't started it yet because I've been busy with something else but for example our notification protocol is using the now-deprecated functionality that allows `ConnectionHandler`s to close connection. Hard to say yet how much work that is to fix.

Thanks for the `libp2p-swarm-test` tip, our tests definitely could use some work and now we've had to comment them out because it was too much work to rework them in every upgrade. But I think for our notification protocol we'd be interested in tracking stream upgrades as well (or at least their failures) because they're related to external code that tracks those stream states and handles connectivity.

>Try to expose yourself as little as possible to the APIs of NetworkBehaviour and ConnectionHandler. Don't write (many) tests against it.

This has definitely been on our TODO list but since practically our entire libp2p-facing code has been written by manually implementing `NetworkBehaviour` for each protocol and notification protocol makes heavy use of both `NetworkBehaviour`/`ConnectionHandler`, it'd essentially be a rewrite of that code, unless I've misunderstood you.

Thanks for your comments

Reply:
> > Try to expose yourself as little as possible to the APIs of NetworkBehaviour and ConnectionHandler. Don't write (many) tests against it.
> 
> This has definitely been on our TODO list but since practically our entire libp2p-facing code has been written by manually implementing `NetworkBehaviour` for each protocol and notification protocol makes heavy use of both `NetworkBehaviour`/`ConnectionHandler`, it'd essentially be a rewrite of that code, unless I've misunderstood you.

Fine to implement `NetworkBehaviour` and `ConnectionHandler`. That is still very much the way to implement custom protocols, e.g. your notification protocol. I think what @thomaseizinger meant here is to not write tests against your `NetworkBehaviour` and `ConnectionHandler` implementations directly, but instead test them through the `libp2p-swarm-test` crate, which boils down to a bunch of helpers to test through the `Swarm` interface.

Reply:
> > I'd be curious to learn what was particularly difficult about that so we can make better decisions in the future.
> 
> I haven't started it yet because I've been busy with something else but for example our notification protocol is using the now-deprecated functionality that allows `ConnectionHandler`s to close connection. Hard to say yet how much work that is to fix.

For the record, you can still force close connections via `Swarm::close_connection` and `ToSwarm::CloseConnection`. The simplest transition would be to - wherever you currently have `ConnectionHandlerEvent::Close` - send a message to the behaviour with `ConnectionHandlerEvent::NotifyBehaviour` to close a particular connection. That is nearly a 1-to-1 transition with the differnce that the connection will be polled for a tiny bit longer whereas `ConnectionHandlerEvent::Close` was an instant hard-stop.

With `v0.53`, `rust-libp2p` itself no longer closes any connections but entirely relies on the keep-alive algorithm which will close connections once they are idle. We consider that to be better because it means protocols can complete gracefully, it allowed us to remove separate type parameters from the APIs and makes types like `FromSwarm` `Clone` and [maybe `Copy`](https://github.com/libp2p/rust-libp2p/pull/4825)!

To transition, I'd recommend to set `ConnectionHandler::Error` to `Void` in a first step. That forces you to move away from emitting `Close` and should make the next upgrade fairly effortless :)